dgl_sparse/src/elemenwise_op.cc

```diff
@@ -32,5 +34,117 @@ c10::intrusive_ptr<SparseMatrix> SpSpAdd(
+    torch::Tensor rhs_val) {
+  c10::intrusive_ptr<SparseMatrix> lhs_intersect_rhs, rhs_intersect_lhs;
+  torch::Tensor lhs_indices, rhs_indices;
+  std::tie(lhs_intersect_rhs, lhs_indices) =
+      SparseMatrixIntersection(lhs_mat, lhs_val, rhs_mat);
+  std::tie(rhs_intersect_lhs, rhs_indices) =
+      SparseMatrixIntersection(rhs_mat, rhs_val, lhs_intersect_rhs);
+  auto ret_mat = SparseMatrix::ValLike(
+      lhs_intersect_rhs,
+      lhs_intersect_rhs->value() * rhs_intersect_lhs->value());
+
+  ctx->saved_data["lhs_require_grad"] = lhs_val.requires_grad();
+  ctx->saved_data["rhs_require_grad"] = rhs_val.requires_grad();
+  if (lhs_val.requires_grad()) {
+    ctx->saved_data["lhs_val_shape"] = lhs_val.sizes().vec();
+    ctx->saved_data["rhs_intersect_lhs"] = rhs_intersect_lhs;
+    ctx->saved_data["lhs_indices"] = lhs_indices;
+  }
+  if (rhs_val.requires_grad()) {
+    ctx->saved_data["rhs_val_shape"] = rhs_val.sizes().vec();
+    ctx->saved_data["lhs_intersect_rhs"] = lhs_intersect_rhs;
+    ctx->saved_data["rhs_indices"] = rhs_indices;
+  }
+  return {ret_mat->Indices(), ret_mat->value()};
+}
+
+tensor_list SpSpMulAutoGrad::backward(
+    AutogradContext* ctx, tensor_list grad_outputs) {
+  torch::Tensor lhs_val_grad, rhs_val_grad;
+  auto output_grad = grad_outputs[1];
+  if (ctx->saved_data["lhs_require_grad"].toBool()) {
+    auto rhs_intersect_lhs =
+        ctx->saved_data["rhs_intersect_lhs"].toCustomClass<SparseMatrix>();
+    const auto& lhs_val_shape = ctx->saved_data["lhs_val_shape"].toIntVector();
+    auto lhs_indices = ctx->saved_data["lhs_indices"].toTensor();
+    lhs_val_grad = torch::zeros(lhs_val_shape, output_grad.options());
+    auto intersect_grad = rhs_intersect_lhs->value() * output_grad;
+    lhs_val_grad.index_put_({lhs_indices}, intersect_grad);
+  }
+  if (ctx->saved_data["rhs_require_grad"].toBool()) {
+    auto lhs_intersect_rhs =
+        ctx->saved_data["lhs_intersect_rhs"].toCustomClass<SparseMatrix>();
+    const auto& rhs_val_shape = ctx->saved_data["rhs_val_shape"].toIntVector();
+    auto rhs_indices = ctx->saved_data["rhs_indices"].toTensor();
+    rhs_val_grad = torch::zeros(rhs_val_shape, output_grad.options());
+    auto intersect_grad = lhs_intersect_rhs->value() * output_grad;
+    rhs_val_grad.index_put_({rhs_indices}, intersect_grad);
+  }
+  return {torch::Tensor(), lhs_val_grad, torch::Tensor(), rhs_val_grad};
+}
+
+c10::intrusive_ptr<SparseMatrix> SpSpMul(
+    const c10::intrusive_ptr<SparseMatrix>& lhs_mat,
+    const c10::intrusive_ptr<SparseMatrix>& rhs_mat) {
+  ElementwiseOpSanityCheck(lhs_mat, rhs_mat);
+  if (lhs_mat->HasDiag() && rhs_mat->HasDiag()) {
+    return SparseMatrix::FromDiagPointer(
+        lhs_mat->DiagPtr(), lhs_mat->value() * rhs_mat->value(),
+        lhs_mat->shape());
+  }
+  TORCH_CHECK(
+      !lhs_mat->HasDuplicate() && !rhs_mat->HasDuplicate(),
+      "Only support SpSpMul on sparse matrices without duplicate values")
```

Comment:
The `HasDuplicate()` is costly as it requires sorting and a linear scan (on GPU, it will incur CPU-GPU synchronization). I understand that SpSpMul shall not support matrices with duplicate entries. My question is what is the general best practice to handle those cases? I see three options:
1. Use a heavy check like the code here.
2. Try to design a light check.
3. Say this is an undefined behavior but make sure the operation will not crash.

cc @frozenbugs